pingcap/tidb-operator
# Add Volume Replacement feature with support for pd, tidb, tikv

### What problem does this PR solve?

Implement a generic feature to make volume changes to tidb via replacement.
Detects volume changes (storageclass change, size change, or number of volumes change)
Notably change in number of volumes such as adding a separate TiKV raft disk is poorly handled by operator today.

Deletes replicas one by one and replaces with new disks that use the new volume parameters in a graceful way similar to how scale in is implemented for components today. (e.g for TiKV it deletes stores and waits for transition).
Adds one spare replica (for PD & TiKV) for handling load while the current one is being deleted.

PR Implements support for TiDB, PD, TiKV components.

More details of problem described in https://docs.google.com/document/d/1WpAYVcF1fXpcuVH6mFlz9v0fbLG3EA9r2gsDgFdqW8E/edit#heading=h.sv3296iop7bh

### What is changed and how does it work?

Detailed List of changes:
* Refactor some volume comparing utils from `pod_vol_modifier.go` & `pvc_modifier.go` in to `vol_compare_utils.go` for reuse in this feature as well as some new functions to compare volume changes in pods, etc.
* Explicitly detect extra volumes (or missing volumes) when comparing statefulset and throw an error as unsupported for pvc_modifier's use case. (Which was previously ignored silently)
* Add a `tidb.pingcap.com/replace-disk` annotation that can be used to trigger a pvc replace implemented in `pkg/controller/tidbcluster/pod_control.go`.
  - TiKV
    1. Get Store ID for Pod via Status/Label. Ensure label is set with store id, this is important to avoid races during scale down later.
    2. Ensure all TiKVs are currently ready and stable.
    3. If store is Up, delete it via PD and wait for it to become Tombstone
    4. Delete all PVCs associated with pod, then delete the pod itself.
    5. StatefulSet will automatically. recreate a new replacement which will come up on its own.
  - TiDB: If other dependent components & TiDB instances are stable, go ahead and delete the pod's PVCs & the pod itself.
  - PD:
    1. Get Member ID for this PD Pod via label
    2. If this pod is leader, transfer leader to next available member.
    3. If this pod is still part of PD members & all PD's are ready, delete it as a member
    4. Delete PVCs & the Pod, let statefulset recreate it.
* Add a VolReplaceInProgress Status field to tidbcluster definition, for TiKV, TiDB & PD components. + supporting functions for it in component_status.go
* Add a new PvcReplacer component that orchestrates new feature with two functions `UpdateStatus()` & `Sync()`
  - In `tidb_cluster_control.go` `UpdateStatus()` is run before component member manager syncs, and `Sync()` is run after.
  - `UpdateStatus()` is responsible for detecting storage volume changes and sets the `VolReplaceInProgress` status field for the respective component.
  - Each supported component's member manager Sync() does the following when a `VolReplaceInProgress` is set:
    + For TiKV and PD an extra desired replica is configured, which triggers ScaleOut
    + The StatefulSet's update strategy is set to OnDelete, so that config+volume changes can be rolled out atomically (by deleting pod with annotation)
     + Any actual config changes to pod spec is suppressed from being updated by member manager itself (bypassing it's upgrade routines)
  - PvcReplacer's `Sync()` waits until scaling is complete, then deletes statefulset orphaning pods & waits for member manager to recreate it (with new config), then applies the `tidb.pingcap.com/replace-disk` annotation to each pod one by one.
*  Add a `tidb.pingcap.com/tikv-no-active-store-since` annotation for TiKV pods (set by pkg/controller/pod_control.go) when a TiKV pod has a valid store id in label but no active stores from PD (via status tikv stores fields). Used to detect Tombstone stores that have been GC'ed too soon (due to a potential PD early GC race when stores are deleted immediately after creation)
* Handle an edge case in TiKV scale in where a pod's store id can be missing from PD and also not available as tombstone (due to PD GC race). Use no-active-store-since annotation above to assume tombstone'd after some waiting time and scale it in.
* Move `MetaManager.Sync()` in `tidb_cluster_control.go` which sets tikv store labels to run before component member manager Sync's so that label is updated, even if member managers error.



### Code changes


### Tests



### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support a generic disk replacement mechanism for making pvc changes behind a new feature (default off) for TiDB, TiKV & PD components.
```


### Manual Test

Here's a procedure to manually test this feature on a test cluster:

Enable feature by setting `features: [VolumeReplacing=true]` in `charts/tidb-operator/values.yaml`
Start a local operator using `./hack/local-up-operator.sh`

Apply the basic tidb cluster but with 3 tikv replicas:
```
cp examples/basic/tidb-cluster.yaml /tmp/basic.yaml
patch /tmp/basic.yaml << EOF
33c33
<     replicas: 1
---
>     replicas: 3
EOF
kubectl --context kind-kind apply -f /tmp/basic.yaml
```

Once cluster is up, apply an update adding a second volume with a config change to be rolled out atomically:
```
cp /tmp/basic.yaml /tmp/twodisk.yaml
patch /tmp/twodisk.yaml << EOF
47a48,53
>       raft-engine:
>         dir: /var/lib/raftlog/raftdb
>     storageVolumes:
>       - name: raftdb-volume
>         storageSize: 1Gi
>         mountPath: /var/lib/raftlog
EOF
kubectl --context kind-kind apply -f /tmp/twodisk.yaml
```

Verify that new temporary tikv replica is created, new volumes created by deleting & replace, and spare replica cleaned up.
Takes about ~10min for test cluster.
